### PR TITLE
Allow users to customize the footer of a login theme

### LINKED ISSUE #31390: Allow custom login themes to define a footer ftl fragment
### Description

It should be easy to add a custom footer to custom login themes without having to override the `template.ftl`.

### Discussion

_No response_

### Motivation

It is a common requirement for custom themes to show a list of links at the footer of the login pages.
Currently users need to replace the whole `template.ftl` file just to add a custom footer fragment.
This leads to more complex themes and causes additional maintenance effort on upgrades.

### Details

It should be possible to define a custom footer just by declaring a custom `footer.ftl` file in the theme.

With the default keycloak theme we could ship an empty `footer.ftl` file that is then referenced by the main `template.ftl` like this:

```html
<#import "footer.ftl" as footer>
...

<@footer.content/>
```

The "empty" default footer.ftl could look like this:
```
<#-- override this file in your custom theme to declare a custom footer element -->
```

The following is an example for a potential custom footer.ftl file:
```html
<#macro content>
    <footer>
        <ul class="footer-links">
            <li><a href="https://apps.training.test/apps/site/home">Home</a></li>
            <li><a href="https://apps.training.test/apps/site/about">About</a></li>
            <li><a href="https://apps.training.test/apps/site/services">Services</a></li>
            <li><a href="https://apps.training.test/apps/site/contact">Contact</a></li>
        </ul>
    </footer>
</#macro>
```

This could be added to the `base` and `keycloak.v2` theme.

### PR DESCRIPTION
**Pull request by thomasdarimont**
Previously, users had to copy the whole template.ftl file or resort to css/js hacks to add a custom footer.

This PR adds new footer templates to the `base/login` and `keycloak.v2/login` themes to allow users
to configure a footer for the login box.

The new `footer.ftl` template contains a content macro that can be added at the bottom of the "login box".

Adjusted `template.ftl` in base login theme to render the macro from `footer.ftl`

The new templates are empty by default to avoid conflicts with existing themes. 

Fixes #31390

### CODE REVIEW

**Comment by thomasdarimont** [comment]
An example custom login theme with a custom footer could look like this:

theme.properties:
```
parent=keycloak
import=common/keycloak
styles=css/login.css css/custom-login.css
```

resources/css/custom-login.css
```css
#kc-footer-links {
    text-align: center;
}
#kc-footer-links li{
    display:inline-block;
}
```

`footer.ftl` file:
```html
<#macro content>
    <!-- hello from footer.html -->
    <hr id="footer-divider">
    <ul id="kc-footer-links">
        <li><a href="https://apps.training.test/apps/site#home">Home</a></li>
        <li><a href="https://apps.training.test/apps/site#about">About</a></li>
        <li><a href="https://apps.training.test/apps/site#services">Services</a></li>
        <li><a href="https://apps.training.test/apps/site#contact">Contact</a></li>
    </ul>
</#macro>
```

Gives the following result:
image

With Identity Providers:
image

This could also be added to the keycloak.v2 login theme.

**Comment by cturkalj** [comment]
Great idea!
After looking at the result, would it be possible to also add a (page) footer at the bottom of the page (instead of the "login form")?
There you could add other things (like a copyright notice) without making the login form look overcrowded.

**Comment by thomasdarimont** [comment]
@cturkalj I thought about that too. How about the following:

`login-footer.ftl` contains a macro that's rendered at the end of the "login box" (like my current footer.ftl above).
`login.ftl` contains a macro that's rendered at the end of the page as you suggested.

Both macros would be empty by default.

Would that work for you?

**Review by edewit — DISMISSED** [review]

**Comment by thomasdarimont** [comment]
@edewit thanks for your review! What do you think about the suggestions from @cturkalj?

**Comment by cturkalj** [comment]
@thomasdarimont:
Perfect, thank you!

**Comment by thomasdarimont** [comment]
I did a quick example for the suggested extension:
https://gist.github.com/thomasdarimont/8021ef4bf715aae2cb28ca2ca3151ab7#file-template-ftl-L169

File: themes/src/main/resources/theme/base/login/template.ftl
Comment (jonkoops) [inline themes/src/main/resources/theme/base/login/template.ftl:167]: Since this element is empty, perhaps we can simply omit it the tag for now? If folks want to add a footer element they can put it in the footer template.

Comment (thomasdarimont) [reply themes/src/main/resources/theme/base/login/template.ftl:167]: Thanks for pointing that out @jonkoops , yes we can give the user complete control of the fragment then.
Would you mind taking a look at the gist in https://github.com/keycloak/keycloak/pull/31391#issuecomment-2235836319 ?

I updated the gist and included your suggestion.

**Comment by thomasdarimont** [comment]
I updated the PR based on the suggestions. I now use empty footer.ftl and login-footer.ftl templates by default. 
Users have full control over the respective contents.

Allthough I think the footer.ftl will rarely be used as the login-footer.ftl is sufficient for many scenarios already. Users can use the footer.ftl to include javascript add the very bottom of the page if necessary.

**Review by jonkoops — DISMISSED** [review]
LGTM!

**Comment by thomasdarimont** [comment]
Added note to release notes

**Comment by ahus1** [comment]
@thomasdarimont - there is a conflict with the release notes. Please rebase and resolve.

**Comment by thomasdarimont** [comment]
@ahus1 Thanks for the review. I rebased the PR and resolved the conflict.

I also added support for the custom footer(s) to the login.v2 theme which I learned will become the default login theme from now on.

However, there are some issues with the login theme v2 which I wrote on the maintainers mailing list. So I'd recommend to wait for feedback from the maintainers before merging this PR.

**Comment by thomasdarimont** [comment]
I also tested this with the keycloak.v2 login theme.

Theme minimal.v2/login, see: https://gist.github.com/thomasdarimont/5fe725319a5fc3c8242669deddd934a8

Custom theme based on Login V2 with Footer 
image

Custom theme based on Login V2 with Identity Providers and Footer 
image

**Comment by thomasdarimont** [comment]
@cturkalj I experimented a bit further with the footer customizations and I think it would be sufficient to only have one `footer.ftl` file that we add at the bottom of the login box.  Links that should be shown at the bottom of the page can be moved there via CSS. This would make the templating easier, as the current login page styling of the keycloak and keycloak.v2 theme does not support adding links outside of the "login box" anyways.

Can you think of something that would not be possible if we only had one footer.ftl ( the login-footer.ftl in the current examples)?

File: docs/documentation/release_notes/topics/26_0_0.adoc
Comment (jonkoops) [inline docs/documentation/release_notes/topics/26_0_0.adoc:53]: Can we also add this information to the theming documentation?

Comment (thomasdarimont) [reply docs/documentation/release_notes/topics/26_0_0.adoc:53]: Yes!

**Comment by cturkalj** [comment]
@thomasdarimont:
I was able to use the login-footer.ftl only and have the footer at the bottom of the page with css (thanks for the hint). 
Besides that, I currently don't have in other use cases in mind.

Now that we have just one (login)-footer.ftl left, should it be renamed it back to footer.ftl?

**Comment by thomasdarimont** [comment]
I revised the PR to only contain 1 additional template file -> `footer.ftl`

- Added section to the release-notes
- Added section to the upgrading guide
- Added section to the themes section in the server development docs

File: docs/documentation/release_notes/topics/26_0_0.adoc
Comment (thomasdarimont) [reply docs/documentation/release_notes/topics/26_0_0.adoc]: Just updated the docs, please check :)

**Review by jonkoops — APPROVED** [review]

**Review by edewit — APPROVED** [review]

**Review by ahus1 — APPROVED** [review]
Approving based on @jonkoops' and @edewit's review. Thanks!

**Pull request merged**
